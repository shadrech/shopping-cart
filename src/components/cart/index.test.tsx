import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render, screen } from '~test-utils';
import { Cart } from '.';

describe('<Cart />', () => {

  it('renders and displays all items', async () => {
    const { asFragment } = render(<Cart />);
    expect(await screen.findByText(/Toilet Roll/i)).toBeInTheDocument();
    expect(screen.getAllByTestId('cartItem')).toHaveLength(3);
    expect(asFragment()).toMatchSnapshot();
  })

  it('correctly updates item quantity when input changed', async () => {
    render(<Cart />);
    await screen.findAllByText('Pasta'); // wait for page to load

    const element1 = screen.getByText('Toilet Roll').parentElement as Element;
    const element2 = screen.getByText('Pasta').parentElement as Element;
    const element3 = screen.getByText('Eggs').parentElement as Element;

    fireEvent.change(element1.querySelector('input') as Element, {
      target: { value: '2' }
    });
    fireEvent.change(element2.querySelector('input') as Element, {
      target: { value: '1' }
    });
    fireEvent.change(element3.querySelector('input') as Element, {
      target: { value: '1' }
    });

    expect(screen.getByTestId('cartTotal')).toHaveTextContent('$3.39');
  })

  it('correctly updates when clear button clicked', async () => {
    render(<Cart />);
    await screen.findAllByText('Pasta');

    screen.getAllByTestId('cartItem').forEach((item) => {
      fireEvent.change(item.querySelector('input') as Element, {
        target: { value: '1' }
      });
    });

    expect(screen.getByTestId('cartTotal')).toHaveTextContent('$2.09');

    fireEvent.click(screen.getByTestId('clearButton'));

    screen.getAllByTestId('cartItem').forEach((item) => {
      expect(item.querySelector<HTMLInputElement>('input')?.value).toEqual('0');
    });
    expect(screen.getByTestId('cartTotal')).toHaveTextContent('$0.00');
  })

  it('displays error when quantity greater than 3 is typed', async () => {
    render(<Cart />);
    const element = (await screen.findByText('Pasta')).parentElement as HTMLElement;

    fireEvent.change(element.querySelector('input') as Element, {
      target: { value: '6' }
    });
    
    expect(element.querySelector('[data-testid="errorMessage"]')).toHaveTextContent('Max 3 items allowed');
    expect(element.querySelector<HTMLInputElement>('input')?.value).toEqual('0');
  })

  it('correctly deletes an item and updates total', async () => {
    render(<Cart />);
    const element = (await screen.findByText('Pasta')).parentElement as HTMLElement;

    fireEvent.change(element.querySelector('input') as Element, {
      target: { value: '2' }
    });
    fireEvent.click(element.querySelector('button') as Element);

    expect(screen.getAllByTestId('cartItem')).toHaveLength(2);
    expect(screen.queryByText('Pasta')).not.toBeInTheDocument();
    expect(screen.getByTestId('cartTotal')).toHaveTextContent('$0.00');
  })

});
