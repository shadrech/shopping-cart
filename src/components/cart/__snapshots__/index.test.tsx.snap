// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders and displays all items 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ bOjPMT"
  >
    <section>
      <div
        class="sc-hKgILt TWQQb"
        data-testid="cartItem"
      >
        <p
          class="sc-gKsewC hyWlBL"
        >
          Pasta
        </p>
        <input
          class="sc-eCssSg brXOyM"
          type="number"
          value="0"
        />
        <p
          class="sc-iBPRYJ frAHyf"
        >
          $0.58
        </p>
        <button
          class="sc-fubCfw eJBkJK"
        >
          x
        </button>
      </div>
      <div
        class="sc-hKgILt TWQQb"
        data-testid="cartItem"
      >
        <p
          class="sc-gKsewC hyWlBL"
        >
          Eggs
        </p>
        <input
          class="sc-eCssSg brXOyM"
          type="number"
          value="0"
        />
        <p
          class="sc-iBPRYJ frAHyf"
        >
          $0.21
        </p>
        <button
          class="sc-fubCfw eJBkJK"
        >
          x
        </button>
      </div>
      <div
        class="sc-hKgILt TWQQb"
        data-testid="cartItem"
      >
        <p
          class="sc-gKsewC hyWlBL"
        >
          Toilet Roll
        </p>
        <input
          class="sc-eCssSg brXOyM"
          type="number"
          value="0"
        />
        <p
          class="sc-iBPRYJ frAHyf"
        >
          $1.30
        </p>
        <button
          class="sc-fubCfw eJBkJK"
        >
          x
        </button>
      </div>
    </section>
    <section
      class="sc-gsTCUz eHqvDi"
    >
      <p
        class="sc-jSgupP khyoCi"
        data-testid="cartTotal"
      >
        $0.00
      </p>
      <button
        class="sc-fubCfw eJBkJK"
        data-testid="clearButton"
      >
        Clear
      </button>
      <button
        class="sc-pFZIQ jOMUqo"
      >
        check out &gt;
      </button>
    </section>
  </div>
  ;
</DocumentFragment>
`;
